letsencrypt/models/letsencrypt.py

```diff
@@ -0,0 +1,157 @@
+# -*- coding: utf-8 -*-
+# © 2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+import os
+import logging
+import urllib2
+import urlparse
+import subprocess
+import tempfile
+from openerp import _, api, models, exceptions
+from openerp.tools.config import _get_default_datadir
+from openerp.tools.misc import file_open
+
+
+DEFAULT_KEY_LENGTH = 4096
+_logger = logging.getLogger(__name__)
+
+
+class Letsencrypt(models.AbstractModel):
+    _name = 'letsencrypt'
+    _description = 'Abstract model providing functions for letsencrypt'
+
+    @api.model
+    def get_data_dir(self):
+        return os.path.join(_get_default_datadir(), 'letsencrypt')
+
+    @api.model
+    def call_cmdline(self, cmdline, loglevel=logging.INFO,
+                     raise_on_result=True):
+        process = subprocess.Popen(
+            cmdline, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        stdout, stderr = process.communicate()
+        if stderr:
+            _logger.log(loglevel, stderr)
+        if stdout:
+            _logger.log(loglevel, stdout)
+
+        if process.returncode:
+            raise exceptions.Warning(
+                _('Error calling %s: %d') % (cmdline[0], process.returncode),
+                ' '.join(cmdline),
+            )
+
+        return process.returncode
+
+    @api.model
+    def generate_account_key(self):
+        data_dir = self.get_data_dir()
+        if not os.path.isdir(data_dir):
+            os.makedirs(data_dir)
+        account_key = os.path.join(data_dir, 'account.key')
+        if not os.path.isfile(account_key):
+            _logger.info('generating rsa account key')
+            self.call_cmdline([
+                'openssl', 'genrsa', '-out', account_key,
+                str(DEFAULT_KEY_LENGTH),
+            ])
+        assert os.path.isfile(account_key), 'failed to create rsa key'
```

Comment:
to me, this is a clear case of can't happen in the list of places to consider putting assertions in https://wiki.python.org/moin/UsingAssertionsEffectively